hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/HiveSyncTool.java

```diff
@@ -307,14 +315,161 @@ private boolean syncSchema(String tableName, boolean tableExists, boolean useRea
+    if (partitionKeys.size() != partitionValues.size()) {
+      throw new HoodieHiveSyncException("Partition key and values should be same length"
+          + ", but got partitionKey: " + partitionKeys + " with values: " + partitionValues);
+    }
+
+    StringBuilder filterBuilder = new StringBuilder();
+    for (int i = 0; i < partitionValues.size(); i++) {
+      String key = partitionKeys.get(i);
+      combineFilter(filterBuilder, "AND",
+          generateEqualFilter(key, keyWithTypes.get(key), partitionValues.get(i)));
+    }
+
+    return filterBuilder.toString();
+  }
+
+  protected String generateWrittenPartitionsFilter(String tableName,
+                                                   List<String> partitionKeys,
+                                                   List<List<String>> partitionVals) {
+    // Hive store columns to lowercase, so we need to map partitions to lowercase to avoid any mismatch.
+    List<String> normalizedPartitionKeys = partitionKeys.stream()
+        .map(String::toLowerCase)
+        .collect(Collectors.toList());
+    List<String> partitionTypes = syncClient.getMetastoreFieldSchemas(tableName)
+        .stream()
+        .filter(f -> normalizedPartitionKeys.contains(f.getName()))
+        .map(FieldSchema::getType)
+        .collect(Collectors.toList());
+
+    if (partitionTypes.size() == 0) {
+      throw new HoodieHiveSyncException("Cannot get partition types from SyncClient, maybe "
+          + "table schema is not synced");
+    }
+
+    Map<String, String> keyWithTypes = CollectionUtils.zipToMap(normalizedPartitionKeys, partitionTypes);
+
+    StringBuilder filterBuilder = new StringBuilder();
+    for (int i = 0; i < partitionVals.size(); i++) {
+      combineFilter(filterBuilder, "OR",
+          Option.of(visitPartition(normalizedPartitionKeys, keyWithTypes, partitionVals.get(i))));
+    }
+    return filterBuilder.toString();
+  }
+
+  /**
+   * Fetch partitions from meta service, will try to push down more filters to avoid fetching
+   * too many unnecessary partitions.
+   */
+  private List<Partition> getTablePartitions(String tableName, List<String> writtenPartitionsSince) {
+    PartitionValueExtractor partitionValueExtractor = ReflectionUtils
+        .loadClass(config.getStringOrDefault(META_SYNC_PARTITION_EXTRACTOR_CLASS));
+    List<String> partitionKeys = config.getSplitStrings(META_SYNC_PARTITION_FIELDS);
+    List<List<String>> partitionVals = writtenPartitionsSince
+        .stream().map(partitionValueExtractor::extractPartitionValuesInPath)
+        .filter(values -> !values.isEmpty())
+        .collect(Collectors.toList());
+
+    if (partitionVals.isEmpty()) {
+      // No partition is written
+      return Collections.emptyList();
+    }
+
+    int estimateSize = partitionKeys.size() * partitionVals.size();
+    if (estimateSize > config.getIntOrDefault(META_SYNC_FILTER_PUSHDOWN_MAX_SIZE)) {
+      return syncClient.getAllPartitions(tableName);
```

Comment:
here we changed the default behavior for many users where filter size < 1000. We have to ensure the correctness of filter generation by having enough UT coverage. So I'd suggest keep the existing sync all partitions still as default and use filter only when people set some threshold, e.g., default = -1 (not using filter). then people set accordingly to their tables' partition numbers

Reply:
sure